File: services/sidecar/internal/api/ai_invoice_actions.go
package api

import (
	"encoding/json"
	"errors"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/google/uuid"
	"github.com/opc-workspace/opc-sidecar/internal/models"
	"gorm.io/gorm"
)

var aiInvoiceActions = []string{"invoice.create", "invoice.update", "invoice.mark_sent", "invoice.mark_viewed", "invoice.mark_paid", "invoice.mark_overdue", "invoice.delete", "invoice.generate_pdf"}

func addAIInvoiceActionSchema(properties map[string]any) {
	action := properties["action"].(map[string]any)
	for _, name := range aiInvoiceActions {
		action["enum"] = append(action["enum"].([]any), name)
	}
	properties["invoice_id"] = map[string]any{"type": "string", "format": "uuid", "description": "Actual invoice ID from workspace_finance; existing invoices also require expected_version"}
	changes := properties["changes"].(map[string]any)
	changes["description"] = changes["description"].(string) + " Invoice proposals require finance + invoice_actions, NOT finance_actions/work/actions. Create requires client_id, amount_minor, currency, issue_date, due_date; optional project_id and full user-supplied notes. Update those fields only while draft. No manual invoice number/status. mark_sent/viewed/overdue, delete and generate_pdf take empty changes; mark_paid requires actual paid_date only (viewed/overdue invoice, full amount received, never intent). All require separate HUMAN invoice consent. delete requires additional HUMAN deletion consent: draft only, no linked ledger, permanently removes invoice and its stored PDF, no undo, approval/audit history retained. mark_overdue requires past due_date and can trigger already-enabled local automation tasks. generate_pdf requires separate HUMAN PDF consent; renders the full invoice locally, replaces existing PDF with no undo, no invoice state change. Human card/download binds the actual asset; no bytes/paths to model. No external sending/payment or CSV export."
	f := changes["properties"].(map[string]any)
	f["issue_date"] = map[string]any{"type": "string", "description": "Explicit invoice issue date YYYY-MM-DD; create year 2000-9999"}
	// Shared fields must retain Task timestamps/clearing and Project dates.
	f["due_date"].(map[string]any)["description"] = "Task: RFC3339 timestamp with explicit time zone. Project: YYYY-MM-DD. Null clears Task/Project only. Invoice: required YYYY-MM-DD >= issue_date, never null."
	f["client_id"].(map[string]any)["description"] = "Project customer association and followup/activity require clients scope. Ledger optional customer uses finance+finance_actions; Invoice required non-null customer uses finance+invoice_actions. Use actual UUIDs; financial grants do not grant customer-profile reads."
	f["paid_date"] = map[string]any{"type": "string", "description": "Actual full payment date YYYY-MM-DD, between issue_date and today; ask user if unknown"}
}

func parseAIInvoiceAction(input aiWorkspaceAction, fields, raw map[string]json.RawMessage) (aiWorkspaceAction, error) {
	known := false
	for _, name := range aiInvoiceActions {
		if input.Action == name {
			known = true
		}
	}
	if !known {
		return input, errors.New("unsupported invoice action")
	}
	for key := range raw {
		if key != "action" && key != "invoice_id" && key != "expected_version" && key != "changes" {
			return input, errors.New("invoice actions only accept their invoice target")
		}
	}
	creating, updating := input.Action == "invoice.create", input.Action == "invoice.update"
	if creating {
		if _, ok := raw["invoice_id"]; ok {
			return input, errors.New("create has no invoice target")
		}
		if _, ok := raw["expected_version"]; ok {
			return input, errors.New("create has no invoice version")
		}
	} else if id, err := uuid.Parse(input.InvoiceID); err != nil || id.String() != input.InvoiceID || input.ExpectedVersion < 1 {
		return input, errors.New("read actual invoice ID/version first")
	}
	canonical := map[string]any{}
	for key, value := range fields {
		if !creating && !updating && (input.Action != "invoice.mark_paid" || key != "paid_date") {
			return input, errors.New("invoice status commands only accept paid_date for mark_paid")
		}
		if (creating || updating) && key == "paid_date" {
			return input, errors.New("paid_date requires mark_paid")
		}
		if key == "amount_minor" {
			var n int64
			if json.Unmarshal(value, &n) != nil || validateInvoiceAmount(n) != nil {
				return input, errors.New("invalid invoice amount_minor")
			}
			canonical[key] = n
			continue
		}
		if key == "project_id" && string(value) == "null" {
			canonical[key] = nil
			continue
		}
		var s string
		if string(value) == "null" || json.Unmarshal(value, &s) != nil {
			return input, errors.New("invalid invoice field type")
		}
		switch key {
		case "client_id", "project_id":
			if id, err := uuid.Parse(s); err != nil || id.String() != s {
				return input, errors.New("canonical invoice association ID required")
			}
		case "currency":
			if n, err := normalizeInvoiceCurrency(s); err != nil || s != n {
				return input, errors.New("explicit uppercase invoice currency required")
			}
		case "issue_date", "due_date", "paid_date":
			if !validDate(s) {
				return input, errors.New("invoice dates must be explicit YYYY-MM-DD")
			}
		case "notes":
			if utf8.RuneCountInString(s) > 10000 {
				return input, errors.New("invoice notes exceed 10000 characters")
			}
		default:
			return input, errors.New("unsupported invoice field")
		}
		canonical[key] = s
	}
	if creating {
		for _, key := range []string{"client_id", "amount_minor", "currency", "issue_date", "due_date"} {
			if _, ok := canonical[key]; !ok {
				return input, errors.New("invoice create requires explicit " + key)
			}
		}
		if canonical["issue_date"].(string) < "2000-01-01" {
			return input, errors.New("invoice numbering requires an issue year between 2000 and 9999")
		}
	}
	if updating && len(canonical) == 0 {
		return input, errors.New("invoice update requires editable fields")
	}
	if input.Action == "invoice.mark_paid" && canonical["paid_date"] == nil {
		return input, errors.New("actual paid_date required")
	}
	input.Changes, _ = json.Marshal(canonical)
	return input, nil
}

func aiInvoiceFields(tx *gorm.DB, invoice models.Invoice) (map[string]any, error) {
	var client struct{ Name string }
	if err := tx.Table("clients").Select("name").Where("id=?", invoice.ClientID).Take(&client).Error; err != nil {
		return nil, err
	}
	var projectName *string
	if invoice.ProjectID != nil {
		var project struct{ Name string }
		if err := tx.Table("projects").Select("name").Where("id=?", *invoice.ProjectID).Take(&project).Error; err != nil {
			return nil, err
		}
		projectName = &project.Name
	}
	var number *string
	if invoice.InvoiceNumber != "" {
		number = &invoice.InvoiceNumber
	}
	return map[string]any{"invoice_number": number, "client_id": invoice.ClientID, "client_name": client.Name, "project_id": invoice.ProjectID, "project_name": projectName,
		"amount_minor": invoice.AmountMinor, "currency": invoice.Currency, "issue_date": invoice.IssueDate, "due_date": invoice.DueDate, "paid_date": invoice.PaidDate, "status": invoice.Status, "notes": invoice.Notes}, nil
}

func previewAIInvoiceAction(tx *gorm.DB, input aiWorkspaceAction, now time.Time) (aiActionPreview, error) {
	p := aiActionPreview{Before: map[string]any{}, After: map[string]any{}}
	var invoice models.Invoice
	var err error
	if input.Action == "invoice.create" {
		var request createInvoiceRequest
		_ = json.Unmarshal(input.Changes, &request)
		invoice, err = invoiceFromCreateRequest(request, now)
		if err == nil {
			err = validateInvoiceAssociations(tx, invoice.ClientID, invoice.ProjectID)
		}
		if err != nil {
			return p, err
		}
		p.Label = "新发票草稿（确认时生成编号）"
	} else {
		invoice, err = prepareInvoiceChange(tx, input.InvoiceID, input.ExpectedVersion)
		if err != nil {
			return p, err
		}
		p.Label = invoice.InvoiceNumber
		p.Before, err = aiInvoiceFields(tx, invoice)
		if err != nil {
			return p, err
		}
		if input.Action == "invoice.delete" || input.Action == "invoice.generate_pdf" {
			if input.Action == "invoice.delete" {
				if err := validateInvoiceDeletion(tx, invoice); err != nil {
					return p, err
				}
			}
			asset, exists, err := invoicePDFAssetExists(tx, invoice.ID)
			if err != nil {
				return p, err
			}
			// Bind stable file identity, not transient integrity-check timestamps.
			// These values stay on the human card; never expose paths or bytes.
			for _, key := range []string{"pdf_asset_id", "pdf_file_name", "pdf_size_bytes", "pdf_sha256", "pdf_generated_from_version", "pdf_generated_at"} {
				p.Before[key] = nil
			}
			if exists {
				p.Before["pdf_asset_id"], p.Before["pdf_file_name"] = asset.ID, asset.FileName
				p.Before["pdf_size_bytes"], p.Before["pdf_sha256"] = asset.SizeBytes, asset.SHA256
				p.Before["pdf_generated_from_version"], p.Before["pdf_generated_at"] = asset.GeneratedFromVersion, asset.GeneratedAt
			}
			p.After = map[string]any{"invoice_deleted": true, "pdf_removed": exists}
			if input.Action == "invoice.generate_pdf" {
				p.After = map[string]any{"pdf_generated": true, "pdf_replaced": exists}
			}
			return p, nil
		}
		if input.Action == "invoice.update" {
			if invoice.Status != "draft" {
				return p, newInvoiceRequestError(409, "INVOICE_NOT_DRAFT", "Only draft invoices can be edited")
			}
			var request updateInvoiceRequest
			_ = json.Unmarshal(input.Changes, &request)
			_, invoice, err = invoiceUpdates(invoice, request)
			if err == nil {
				err = validateInvoiceAssociations(tx, invoice.ClientID, invoice.ProjectID)
			}
		} else {
			var request transitionInvoiceRequest
			_ = json.Unmarshal(input.Changes, &request)
			invoice.Status, _, _, err = invoiceTransition(invoice, strings.TrimPrefix(input.Action, "invoice."), request.PaidDate, now)
			if input.Action == "invoice.mark_paid" {
				invoice.PaidDate = request.PaidDate
			}
		}
		if err != nil {
			return p, err
		}
	}
	p.After, err = aiInvoiceFields(tx, invoice)
	return p, err
}

func executeAIInvoiceAction(tx *gorm.DB, input aiWorkspaceAction, requestID string, clock time.Time) (aiActionResult, error) {
	// Dates follow the same local calendar as native Invoice commands. Only
	// persisted timestamps are normalized to UTC, never the validation clock.
	now := clock.UTC().Format(time.RFC3339Nano)
	var err error
	var out invoiceResponse
	switch input.Action {
	case "invoice.create":
		var request createInvoiceRequest
		_ = json.Unmarshal(input.Changes, &request)
		invoice, e := invoiceFromCreateRequest(request, clock)
		if e != nil {
			return aiActionResult{}, e
		}
		out, err = createInvoiceInTransaction(tx, invoice, requestID)
	case "invoice.update":
		var request updateInvoiceRequest
		_ = json.Unmarshal(input.Changes, &request)
		out, err = updateInvoiceInTransaction(tx, input.InvoiceID, input.ExpectedVersion, request, requestID, now)
	default:
		var request transitionInvoiceRequest
		_ = json.Unmarshal(input.Changes, &request)
		out, err = transitionInvoiceInTransaction(tx, input.InvoiceID, input.ExpectedVersion, strings.TrimPrefix(input.Action, "invoice."), request.PaidDate, clock, requestID)
	}
	return aiActionResult{ID: out.ID, Version: out.Version}, err
}

File: apps/web/src/api/aiInvoiceActions.ts
import {
  focusReportDays,
  isWorkspaceIdentity,
} from "../lib/focusReportLocation";
import type { AiActionProposal } from "./aiWorkspaceActions";

export interface AiInvoicePdfResult {
  asset_id: string;
  invoice_id: string;
  file_name: string;
  mime_type: "application/pdf";
  size_bytes: number;
  sha256: string;
  generated_from_version: number;
  generated_at: string;
  integrity_status: "verified";
  integrity_checked_at: string;
}

export function validInvoicePdfResult(p: AiActionProposal): boolean {
  const v = p.invoice_pdf_result;
  if (p.action.action !== "invoice.generate_pdf" || p.status !== "confirmed")
    return v === undefined;
  if (!v || typeof v !== "object" || Array.isArray(v)) return false;
  const keys = [
    "asset_id",
    "invoice_id",
    "file_name",
    "mime_type",
    "size_bytes",
    "sha256",
    "generated_from_version",
    "generated_at",
    "integrity_status",
    "integrity_checked_at",
  ];
  return (
    Object.keys(v).length === keys.length &&
    keys.every((k) => Object.hasOwn(v, k)) &&
    v.asset_id === p.result_id &&
    v.invoice_id === p.action.invoice_id &&
    v.generated_from_version === p.action.expected_version &&
    v.generated_from_version === p.result_version &&
    v.mime_type === "application/pdf" &&
    v.integrity_status === "verified" &&
    v.integrity_checked_at === v.generated_at &&
    validDeletionPdf({
      pdf_asset_id: v.asset_id,
      pdf_file_name: v.file_name,
      pdf_size_bytes: v.size_bytes,
      pdf_sha256: v.sha256,
      pdf_generated_from_version: v.generated_from_version,
      pdf_generated_at: v.generated_at,
    })
  );
}

const editable = [
  "client_id",
  "project_id",
  "amount_minor",
  "currency",
  "issue_date",
  "due_date",
  "notes",
];
const snapshots = [
  ...editable,
  "client_name",
  "project_name",
  "invoice_number",
  "status",
  "paid_date",
];
const pdfSnapshot = [
  "pdf_asset_id",
  "pdf_file_name",
  "pdf_size_bytes",
  "pdf_sha256",
  "pdf_generated_from_version",
  "pdf_generated_at",
];
function validDeletionPdf(
  before: AiActionProposal["preview"]["before"],
): boolean {
  if (before.pdf_asset_id === null)
    return pdfSnapshot.every((k) => before[k] === null);
  return (
    typeof before.pdf_asset_id === "string" &&
    isWorkspaceIdentity(before.pdf_asset_id) &&
    typeof before.pdf_file_name === "string" &&
    before.pdf_file_name.length >= 5 &&
    before.pdf_file_name.length <= 180 &&
    !/[\\/\u0000-\u001f\u007f]/.test(before.pdf_file_name) &&
    before.pdf_file_name.endsWith(".pdf") &&
    typeof before.pdf_sha256 === "string" &&
    /^[a-f0-9]{64}$/.test(before.pdf_sha256) &&
    typeof before.pdf_size_bytes === "number" &&
    Number.isSafeInteger(before.pdf_size_bytes) &&
    before.pdf_size_bytes > 0 &&
    before.pdf_size_bytes <= 50 * 1024 * 1024 &&
    typeof before.pdf_generated_from_version === "number" &&
    Number.isSafeInteger(before.pdf_generated_from_version) &&
    before.pdf_generated_from_version > 0 &&
    typeof before.pdf_generated_at === "string" &&
    Number.isFinite(Date.parse(before.pdf_generated_at))
  );
}
const date = (v: unknown): v is string =>
  typeof v === "string" && focusReportDays(v, v) === 1;
function field(k: string, v: unknown): boolean {
  switch (k) {
    case "client_id":
      return typeof v === "string" && isWorkspaceIdentity(v);
    case "project_id":
      return v === null || (typeof v === "string" && isWorkspaceIdentity(v));
    case "amount_minor":
      return (
        typeof v === "number" &&
        Number.isSafeInteger(v) &&
        v > 0 &&
        v <= 9_000_000_000_000_000
      );
    case "currency":
      return typeof v === "string" && /^[A-Z]{3}$/.test(v);
    case "issue_date":
    case "due_date":
    case "paid_date":
      return date(v);
    case "notes":
      return typeof v === "string" && Array.from(v).length <= 10000;
    default:
      return false;
  }
}
export function validInvoiceProposal({
  action,
  preview,
}: AiActionProposal): boolean {
  const creating = action.action === "invoice.create",
    deleting = action.action === "invoice.delete",
    generating = action.action === "invoice.generate_pdf",
    updating = action.action === "invoice.update",
    paid = action.action === "invoice.mark_paid";
  const keys = creating
    ? ["action", "changes"]
    : ["action", "changes", "invoice_id", "expected_version"];
  if (Object.keys(action).some((k) => !keys.includes(k))) return false;
  const c = action.changes,
    allowed = creating || updating ? editable : paid ? ["paid_date"] : [];
  if (Object.keys(c).some((k) => !allowed.includes(k) || !field(k, c[k])))
    return false;
  if (
    creating &&
    ["client_id", "amount_minor", "currency", "issue_date", "due_date"].some(
      (k) => !Object.hasOwn(c, k),
    )
  )
    return false;
  if ((updating && !Object.keys(c).length) || (paid && !date(c.paid_date)))
    return false;
  if (creating && Object.keys(preview.before).length) return false;
  for (const [s, after] of [
    [preview.before, false],
    [preview.after, true],
  ] as const) {
    if ((creating && !after) || ((deleting || generating) && after)) continue;
    const required =
      deleting || generating ? [...snapshots, ...pdfSnapshot] : snapshots;
    if (
      Object.keys(s).length !== required.length ||
      required.some((k) => !Object.hasOwn(s, k)) ||
      editable.some((k) => !field(k, s[k]))
    )
      return false;
    if (
      String(s.due_date) < String(s.issue_date) ||
      typeof s.client_name !== "string"
    )
      return false;
    if (
      (s.project_id === null) !== (s.project_name === null) ||
      (s.project_id !== null && typeof s.project_name !== "string")
    )
      return false;
    if (
      creating
        ? s.invoice_number !== null
        : typeof s.invoice_number !== "string" || !s.invoice_number
    )
      return false;
    if (
      !["draft", "sent", "viewed", "paid", "overdue"].includes(String(s.status))
    )
      return false;
    if (
      s.status === "paid"
        ? !date(s.paid_date) || s.paid_date < String(s.issue_date)
        : s.paid_date !== null
    )
      return false;
  }
  const a = preview.after,
    b = preview.before;
  if (deleting || generating)
    return (
      (generating || b.status === "draft") &&
      validDeletionPdf(b) &&
      Object.keys(a).length === 2 &&
      a[generating ? "pdf_generated" : "invoice_deleted"] === true &&
      a[generating ? "pdf_replaced" : "pdf_removed"] ===
        (b.pdf_asset_id !== null) &&
      (b.pdf_asset_id === null ||
        Number(b.pdf_generated_from_version) <= Number(action.expected_version))
    );
  if (creating && String(a.issue_date) < "2000-01-01") return false;
  const transitions: Record<string, [string[], string]> = {
    "invoice.update": [["draft"], "draft"],
    "invoice.mark_sent": [["draft"], "sent"],
    "invoice.mark_viewed": [["sent"], "viewed"],
    "invoice.mark_paid": [["viewed", "overdue"], "paid"],
    "invoice.mark_overdue": [["sent", "viewed"], "overdue"],
  };
  const transition = transitions[action.action];
  if (
    creating
      ? a.status !== "draft"
      : !transition ||
        !transition[0].includes(String(b.status)) ||
        a.status !== transition[1]
  )
    return false;
  for (const k of editable) {
    const expected = Object.hasOwn(c, k)
      ? c[k]
      : creating
        ? k === "notes"
          ? ""
          : null
        : b[k];
    if (a[k] !== expected) return false;
  }
  if (a.paid_date !== (paid ? c.paid_date : creating ? null : b.paid_date))
    return false;
  if (!creating && a.invoice_number !== b.invoice_number) return false;
  for (const r of ["client", "project"]) {
    if (
      !creating &&
      a[`${r}_id`] === b[`${r}_id`] &&
      a[`${r}_name`] !== b[`${r}_name`]
    )
      return false;
  }
  return true;
}
